### DISCUSSION TITLE
v12 BUG? Initializing with nodes/edges doesn't apply useEffect/useHandleConnections/useNodesData business logic

### DISCUSSION THREAD

**Discussion by tauren** [xyflow/xyflow#3849] [Q&A] [answered]
I have a simple flow with 2 nodes and one edge between them. `nodeB` is dependent on `nodeA.data.value`. When I manually drag a connection from `NodeA` to `NodeB`, the business logic in `NodeB` pulls `nodeA.data.value` and adds it to `nodeB.data.value`. The node is rendered correctly.

However, using `<ReactFlow edges={edges} nodes={nodes} />` with two preset nodes and one edge between them, the connection is rendered properly and the logged output (see code below) indicates the business logic does get run. But `nodeB.data.value` doesn't exist or render. It's like everything runs except `updateNodeData` didn't execute, but not when I manually drag connections.

The following is a simplified snippet of my code:

```typescript
const CustomNode = ({ id, data }: NodeProps) => {
  // NodeB
  const { updateNodeData } = useReactFlow();
  const parentConnection = useHandleConnections({
    type: 'target',
  })[0];
  const parentData = useNodesData(parentConnection?.source);

  useEffect(() => {
      const result = {};
      result.value = parentData?.value || '';
      console.log(result) // This logs `{ value: "valueA" }` which is what parentData contains
      updateNodeData(id, result);
  }, [id, updateNodeData, parentData]);

  return (
    <div>{data.value}</div>
  );
};
```

If I click the edge and delete it, then manually link it back, then `nodeB.data.value` exists and is rendered.

I've tried adding `useUpdateNodeInternals`, but it seems to have no impact.

```typescript
  const updateNodeInternals = useUpdateNodeInternals();

  useEffect(() => {
      const result = {};
      result.value = parentData?.value || '';
      console.log(result) // This logs `{ value: "valueA" }` which is what parentData contains
      updateNodeData(id, result);
      updateNodeInternals(id);
  }, [id, updateNodeData, parentData]);
```

Any suggestions? Is this a v12 bug, or am I doing something wrong?

**Comment by moklick** [comment]
we will continue with this discussion here: #3865

**Comment by peterkogo** [comment, marked as answer]
@tauren Please use a controlled flow like here https://reactflow.dev/learn/concepts/core-concepts#basic-functionality

If you want to use an uncontrolled flow (without onNodesChange & onEdgesChange) you have to use `defaultNodes` & `defaultEdges` instead of `nodes` & `edges`. (Note that I just found a small bug, so you are better of using a controlled flow)
